### PR TITLE
Introduce `FIELD_ADDR` and use it for TLS statics and instance class fields

### PR DESCRIPTION
**Pull request by SingleAccretion**
Removing two uses of `GT_ADDR` in the process.

We expect a couple diffs in tests where we now fail to hoist a vector constant because we don't resolve struct handles (thus recognize SIMD types) for `FIELD_ADDR`s, as well as one case of new substitution due to smaller tree size. We also expect a minor TP improvement in the optimized case.

### CODE REVIEW

**Comment by SingleAccretion** [comment]
@dotnet/jit-contrib

**Comment by BruceForstall** [comment]
[Diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=61268&view=ms.vss-build-web.run-extensions-tab)

File: src/coreclr/jit/morph.cpp [hunk 1]
```diff
@@ -5089,325 +5147,323 @@ GenTree* Compiler::fgMorphField(GenTree* tree, MorphAddrContext* mac)
...
+    if (addExplicitNullCheck)
+    {
+        JITDUMP("Before explicit null check morphing:\n");
+        DISPTREE(tree);
 
-            if (!objRef->OperIs(GT_LCL_VAR) || lvaIsLocalImplicitlyAccessedByRef(objRef->AsLclVar()->GetLclNum()))
-            {
-                lclNum = fgGetFieldMorphingTemp(tree->AsField());
+        // Create the "comma" subtree.
+        GenTree* asg = nullptr;
```
Comment (SingleAccretion) [inline src/coreclr/jit/morph.cpp:5176]:
```suggestion
        GenTree* asg = nullptr;
        unsigned lclNum;

```

File: src/coreclr/jit/importer.cpp [hunk 2]
```diff
@@ -9553,49 +9559,50 @@ void Compiler::impImportBlockCode(BasicBlock* block)
...
+                            if (varTypeIsStruct(lclTyp))
+                            {
+                                op1 = gtNewObjNode(fieldInfo.structType, op1);
+                                op1->gtFlags |= GTF_IND_NONFAULTING;
+                            }
+                            else
+                            {
+                                op1 = gtNewIndir(lclTyp, op1, GTF_IND_NONFAULTING);
+                                op1->gtFlags |= GTF_GLOB_REF;
+                            }
```
Comment (jakobbotsch) [inline src/coreclr/jit/importer.cpp:9599]: Doesn't this mean we end up with `GT_OBJ` for SIMD indirs?

Comment (SingleAccretion) [reply src/coreclr/jit/importer.cpp:9599]: Yes, but that's consistent with how the rest of importer operates. These will at some point be switched to `IND`s, but in a coordinated effort.

**Comment by jakobbotsch** [comment]
The unreduced BadResult failures in the Fuzzlyn run looks like a known intermittent failure that I have seen before, I don't think it is related to this PR.

File: src/coreclr/jit/morph.cpp [hunk 3]
```diff
@@ -5089,325 +5147,324 @@ GenTree* Compiler::fgMorphField(GenTree* tree, MorphAddrContext* mac)
...
+// "__declspec(thread)" variables. An overview of the underlying native
+// mechanism can be found here: http://www.nynaeve.net/?p=180.
+//
+// Arguments:
+//    tree - The GT_FIELD_ADDR tree
+//
+// Return Value:
+//    The expanded tree - a GT_ADD.
+//
+GenTree* Compiler::fgMorphExpandTlsFieldAddr(GenTree* tree)
```
Comment (jakobbotsch) [inline src/coreclr/jit/morph.cpp:5283]: I assume there's no actual changes here except constructing the address part of this? The diff is quite hard to read.

**Review by jakobbotsch — APPROVED** [review]

File: src/coreclr/jit/morph.cpp (hunk 3 above)
Comment (SingleAccretion) [reply src/coreclr/jit/morph.cpp:5283]: There is one change: `tlsRef = gtNewIndir(TYP_I_IMPL, tlsRef, GTF_IND_NONFAULTING | GTF_IND_INVARIANT);` - for the indir off of the TLS handle.

(Not doing this was making `fgDebugCheckFlags` fail)

**Comment by jakobbotsch** [comment]
The stress failures were #78023

**Pull request merged**